# Iris Flower Classification - ANN Model

## Introduction

The Iris flower classification project applies an Artificial Neural Network (ANN) using TensorFlow/Keras to classify Iris flowers into three species: Setosa, Versicolor, and Virginica. The dataset consists of four flower measurements, which are used to train the model to recognize patterns and accurately predict the species of unseen flower samples.


## Objective

The objective of this Project is to develop an Artificial Neural Network (ANN) model using TensorFlow/Keras to accurately classify Iris flowers into Setosa, Versicolor, and Virginica based on their flower measurements.

In [1]:
!pip install tensorflow

   ---------------------------------------- 0.0/351.2 MB ? eta -:--:--
   ---------------------------------------- 0.0/351.2 MB ? eta -:--:--
   ---------------------------------------- 0.3/351.2 MB ? eta -:--:--
   ---------------------------------------- 0.5/351.2 MB 1.2 MB/s eta 0:04:56
   ---------------------------------------- 0.8/351.2 MB 1.3 MB/s eta 0:04:38
   ---------------------------------------- 1.0/351.2 MB 1.2 MB/s eta 0:04:43
   ---------------------------------------- 1.3/351.2 MB 1.3 MB/s eta 0:04:34
   ---------------------------------------- 1.8/351.2 MB 1.4 MB/s eta 0:04:07
   ---------------------------------------- 2.1/351.2 MB 1.5 MB/s eta 0:03:53
   ---------------------------------------- 2.6/351.2 MB 1.6 MB/s eta 0:03:38
   ---------------------------------------- 3.1/351.2 MB 1.7 MB/s eta 0:03:29
   ---------------------------------------- 3.4/351.2 MB 1.7 MB/s eta 0:03:25
   ---------------------------------------- 3.9/351.2 MB 1.7 MB/s eta 0:03:25
    ---

ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
streamlit 1.51.0 requires protobuf<7,>=3.20, but you have protobuf 7.35.0 which is incompatible.


In [1]:
# Step 1 : Import Libraries

import numpy as np
import pandas as pd

from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder,StandardScaler

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.utils import to_categorical

In [2]:
# Step 2 :Load Dataset

iris = load_iris()

x = iris.data
y = iris.target

# Create DataFrame 
df = pd.DataFrame(x,columns = iris.feature_names)

# Add target column

df['Target'] = y

# Display first Five rows
df.head()



,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm),Target
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0
2,4.7,3.2,1.3,0.2,0
3,4.6,3.1,1.5,0.2,0
4,5.0,3.6,1.4,0.2,0


In [3]:
# Step 3 : Preprocessing

# Check null values 
print(df.isnull().sum())

# Check Duplicate values
print(df.duplicated().sum())

# Dataset Shape 
print(df.shape)
# Dataset Info
print(df.info())


sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
Target               0
dtype: int64
1
(150, 5)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   Target             150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB
None


In [4]:
# Step 4 : Separate input and output

x = df.drop('Target',axis = 1)
y = df['Target']

In [5]:
# Step 5 : Label Encoding

le = LabelEncoder()

y = le.fit_transform(y)

# One-hot-encoding

y_encoded = to_categorical(y)
print(y_encoded[:5])

[[1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]
 [1. 0. 0.]]


In [6]:
# Step 6 : Normalize features
scaler = StandardScaler()

X_scaled = scaler.fit_transform(x)
print(X_scaled[:5])

[[-0.90068117  1.01900435 -1.34022653 -1.3154443 ]
 [-1.14301691 -0.13197948 -1.34022653 -1.3154443 ]
 [-1.38535265  0.32841405 -1.39706395 -1.3154443 ]
 [-1.50652052  0.09821729 -1.2833891  -1.3154443 ]
 [-1.02184904  1.24920112 -1.34022653 -1.3154443 ]]


In [7]:
# Step 7 :Train-test split

X_train,X_test, y_train, y_test = train_test_split(
    X_scaled,
    y_encoded,
    test_size = 0.2,
    random_state = 42
)

print(X_train.shape)
print(X_test.shape)

(120, 4)
(30, 4)


In [8]:
# Step 8: Build the Model

model = Sequential()
# Hidden Layer 1 
model.add(Dense(10,input_shape=(4,),activation = 'relu'))

# Hidden Layer 2 
model.add(Dense(8,activation='relu'))

# Output Layer 
model.add(Dense(3,activation = 'softmax'))

C:\Users\shifa\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [9]:
# Step 9 : Compile The model 

model.compile(
    optimizer = 'adam',
    loss = 'categorical_crossentropy',
    metrics = ['accuracy']
)

In [10]:
# Step 10 : Train the model

history = model.fit(
    X_train,
    y_train,
    epochs = 50,
    batch_size =8,
    validation_split = 0.2
)


Epoch 1/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 3s 56ms/step - accuracy: 0.4792 - loss: 1.0725 - val_accuracy: 0.5000 - val_loss: 1.0466
Epoch 2/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 1s 21ms/step - accuracy: 0.6771 - loss: 1.0451 - val_accuracy: 0.8750 - val_loss: 1.0081
Epoch 3/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.6979 - loss: 1.0164 - val_accuracy: 0.8750 - val_loss: 0.9744
Epoch 4/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 20ms/step - accuracy: 0.7500 - loss: 0.9871 - val_accuracy: 0.8750 - val_loss: 0.9345
Epoch 5/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.7917 - loss: 0.9574 - val_accuracy: 0.8333 - val_loss: 0.8979
Epoch 6/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 18ms/step - accuracy: 0.8125 - loss: 0.9285 - val_accuracy: 0.8333 - val_loss: 0.8591
Epoch 7/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 21ms/step - accuracy: 0.8125 - loss: 0.8974 - val_accuracy: 0.8333 - val_loss: 0.8248
Epoch 8/50
12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 19ms/step - accuracy: 0.8125 - loss: 0.8678 - val_accuracy: 0.8333 - v

In [11]:
# Step : 11 Evaluate the model

loss,accuracy = model.evaluate(X_test, y_test)

print("Loss:",loss)
print("Accuracy",accuracy)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 460ms/step - accuracy: 0.9667 - loss: 0.1666
Loss: 0.16657452285289764
Accuracy 0.9666666388511658


In [12]:
# Step 12: Prediction

y_pred = model.predict(X_test)

# Convert predicted values
y_pred_classes = np.argmax(y_pred, axis = 1)

# Convert avtual values
y_test_classes = np.argmax(y_test, axis = 1)

print("Predicted Values:")
print(y_pred_classes[:10])

print("Actual Values :")
print(y_test_classes[:10])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 158ms/step
Predicted Values:
[1 0 2 1 1 0 1 2 1 1]
Actual Values :
[1 0 2 1 1 0 1 2 1 1]


In [13]:
# Step 13 : Accuracy Score

from sklearn.metrics import accuracy_score

acc = accuracy_score(y_test_classes,y_pred_classes)
print("Accuracy Score:",acc)

Accuracy Score: 0.9666666666666667


In [16]:
# Step 14 : Classification Report

from sklearn.metrics import classification_report

from sklearn.metrics import classification_report

print(classification_report(
    y_test_classes,
    y_pred_classes,
    target_names=["Setosa", "Versicolor", "Virginica"]
))

              precision    recall  f1-score   support

      Setosa       1.00      1.00      1.00        10
  Versicolor       1.00      0.89      0.94         9
   Virginica       0.92      1.00      0.96        11

    accuracy                           0.97        30
   macro avg       0.97      0.96      0.97        30
weighted avg       0.97      0.97      0.97        30



In [18]:
# Step 15 : Save Model

model.save("iris_ann_model.keras")
print("Model saved successfully.")

Model saved successfully.


In [19]:
# Step 16 : Save Scaler
import joblib

joblib.dump(scaler, "scaler.pkl")
print("Scaler saved successfully.")

Scaler saved successfully.


In [22]:
# Step 16 : Test with an unseen flower sample
sample = pd.DataFrame({
    'sepal length (cm)': [5.8],
    'sepal width (cm)': [2.7],
    'petal length (cm)': [5.1],
    'petal width (cm)': [1.9]
})
sample_scaled = scaler.transform(sample)
prediction = model.predict(sample_scaled)
predicted_class = np.argmax(prediction)
flower_names = ["Setosa", "Versicolor", "Virginica"]
print("Predicted Flower Species:", flower_names[predicted_class])

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 76ms/step
Predicted Flower Species: Virginica


## Conclusion

An Artificial Neural Network (ANN) model was successfully developed using TensorFlow/Keras to classify Iris flowers into three species. After preprocessing, training, and evaluation, the model achieved an accuracy of 96.67%, demonstrating its ability to accurately classify unseen flower samples. This project highlights the effectiveness of ANN for multiclass classification tasks.